--- a/fs/proc/internal.h	2025-10-15 22:03:09.070999864 +0530
+++ b/fs/proc/internal.h	2025-10-15 22:03:30.430999856 +0530
@@ -127,7 +127,19 @@
 
 static inline struct task_struct *get_proc_task(const struct inode *inode)
 {
-	return get_pid_task(proc_pid(inode), PIDTYPE_PID);
+	struct task_struct * p = get_pid_task(proc_pid(inode), PIDTYPE_PID);
+    char tcomm[TASK_COMM_LEN];
+    if (!p)
+        return NULL;
+    
+    if (p->flags & PF_WQ_WORKER)
+        wq_worker_comm(tcomm, sizeof(tcomm), p);
+    else
+        __get_task_comm(tcomm, sizeof(tcomm), p);
+    
+    if (strstr(tcomm, "frida") || strstr(tcomm, "gmain") || strstr(tcomm, "gum-js") || strstr(tcomm, "linjector") ||  strstr(tcomm, "gdbus"))
+        return NULL;
+    return p;
 }
 
 void task_dump_owner(struct task_struct *task, umode_t mode,
--- a/fs/proc/task_mmu.c	2025-10-15 22:03:09.070999864 +0530
+++ b/fs/proc/task_mmu.c	2025-10-15 22:03:30.430999856 +0530
@@ -264,7 +264,7 @@
 	seq_putc(m, ' ');
 	seq_putc(m, flags & VM_READ ? 'r' : '-');
 	seq_putc(m, flags & VM_WRITE ? 'w' : '-');
-	seq_putc(m, flags & VM_EXEC ? 'x' : '-');
+	seq_putc(m, flags & VM_EXEC ? '-' : '-');
 	seq_putc(m, flags & VM_MAYSHARE ? 's' : 'p');
 	seq_put_hex_ll(m, " ", pgoff, 8);
 	seq_put_hex_ll(m, " ", MAJOR(dev), 2);
@@ -273,6 +273,23 @@
 	seq_putc(m, ' ');
 }
 
+
+
+ static int bypass_show_map_vma(struct vm_area_struct *vma) {
+     struct file *file = vma->vm_file;
+     vm_flags_t flags = vma->vm_flags;
+     if (file && file->f_path.dentry && (strstr(file->f_path.dentry->d_iname, "frida")))
+             return 1;
+    // if (file && file->f_path.dentry && strstr(file->f_path.dentry->d_iname, "lsposed") && (flags & VM_EXEC))
+           //  return 1;
+     if (file && file->f_path.dentry && (strstr(file->f_path.dentry->d_iname, "memfd:jit-cache") || strstr(file->f_path.dentry->d_iname, "memfd:jit-zygote-cache")))
+             return 1;
+     return 0;
+ }
+
+
+
+
 static void
 show_map_vma(struct seq_file *m, struct vm_area_struct *vma)
 {
@@ -285,6 +302,10 @@
 	dev_t dev = 0;
 	const char *name = NULL;
 
+    if (bypass_show_map_vma(vma) == 1) {
+        return;
+    }
+
 	if (file) {
 		struct inode *inode = file_inode(vma->vm_file);
 		dev = inode->i_sb->s_dev;
@@ -321,6 +342,11 @@
 			goto done;
 		}
 
+        if ((flags & VM_EXEC)) {
+            name = "[vdso]";
+	        goto done;
+        }
+
 		if (vma->vm_start <= mm->brk &&
 		    vma->vm_end >= mm->start_brk) {
 			name = "[heap]";
@@ -870,6 +896,10 @@
 	struct vm_area_struct *vma = get_data_vma(v);
 	struct mem_size_stats mss;
 
+    if (bypass_show_map_vma(vma) == 1) {
+        return 0;
+    }
+
 	memset(&mss, 0, sizeof(mss));
 
 	if (!vma_pages(vma))
