--- a/fs/proc/internal.h	2025-10-16 07:32:43.296004431 +0530
+++ b/fs/proc/internal.h	2025-10-15 22:03:30.430999856 +0530
@@ -127,7 +127,19 @@
 
 static inline struct task_struct *get_proc_task(const struct inode *inode)
 {
-	return get_pid_task(proc_pid(inode), PIDTYPE_PID);
+	struct task_struct * p = get_pid_task(proc_pid(inode), PIDTYPE_PID);
+    char tcomm[TASK_COMM_LEN];
+    if (!p)
+        return NULL;
+    
+    if (p->flags & PF_WQ_WORKER)
+        wq_worker_comm(tcomm, sizeof(tcomm), p);
+    else
+        __get_task_comm(tcomm, sizeof(tcomm), p);
+    
+    if (strstr(tcomm, "frida") || strstr(tcomm, "gmain") || strstr(tcomm, "gum-js") || strstr(tcomm, "linjector") ||  strstr(tcomm, "gdbus"))
+        return NULL;
+    return p;
 }
 
 void task_dump_owner(struct task_struct *task, umode_t mode,
--- a/fs/proc/task_mmu.c	2025-10-16 07:32:43.296004431 +0530
+++ b/fs/proc/task_mmu.c	2025-10-15 22:03:30.430999856 +0530
@@ -349,6 +349,29 @@
 
 static int show_map(struct seq_file *m, void *v)
 {
+
+    struct proc_maps_private *priv = m->private;
+    struct task_struct *target = priv->task;
+    static pid_t last_reader = -1;
+    char buf[256];
+
+    /* Log only once per read — detect first VMA in sequence */
+    if (m->index == 0 && current->pid != last_reader) {
+        snprintf(buf, sizeof(buf),
+            "MAPReader: %s(pid=%d,uid=%d) -> Target: %s(pid=%d)\n",
+            current->comm,
+            current->pid,
+            from_kuid_munged(&init_user_ns, current_cred()->uid),
+            target ? target->comm : "?",
+            target ? task_pid_nr(target) : -1);
+
+        pr_info("%s", buf);
+
+        last_reader = current->pid;
+    }
+
+
+// normal
 	struct vm_area_struct *pad_vma = get_pad_vma(v);
 	struct vm_area_struct *vma = get_data_vma(v);
 
